src/Microsoft.Extensions.Configuration/IncludedConfigurationProvider.cs

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.Configuration
+{
+    public class IncludedConfigurationProvider : ConfigurationProvider
+    {
+        public IncludedConfigurationProvider(IConfiguration source)
+        {
+            if (source == null)
+            {
+                throw new ArgumentNullException(nameof(source));
+            }
+            int pathStart = 0;
+            var section = source as IConfigurationSection;
+            if (section != null)
+            {
+                pathStart = section.Path.Length + 1;
+            }
+            foreach (var child in source.GetChildren())
+            {
+                AddSection(child, pathStart);
+            }
+        }
+
+        private void AddSection(IConfigurationSection section, int pathStart)
```

Comment:
The first time someone hits this provider will be the first time anyone asks for anything on the configuration to which the provider was added. I.e. the only case in which we save any work is if the configuration was built but never used. Not sure how much gain that is. I think @davidfowl was rather referring to not ever copying the data but instead redirecting all requests for keys or values to the underlying configuration.